engine/common/splitter/engine.go

```diff
@@ -0,0 +1,149 @@
+package splitter
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/network"
+	"github.com/rs/zerolog"
+)
+
+type Engine struct {
+	mu      sync.Mutex
+	unit    *engine.Unit               // used to manage concurrency & shutdown
+	log     zerolog.Logger             // used to log relevant actions with context
+	engines map[module.Engine]struct{} // stores registered engines
+	channel network.Channel            // the channel that this splitter listens on
+}
+
+func New(
+	log zerolog.Logger,
+	channel network.Channel,
+) *Engine {
+	e := &Engine{
+		unit:    engine.NewUnit(),
+		log:     log.With().Str("engine", "splitter").Logger(),
+		engines: make(map[module.Engine]struct{}),
+		channel: channel,
+	}
+
+	return e
+}
+
+func (e *Engine) RegisterEngine(engine module.Engine) error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	if _, ok := e.engines[engine]; ok {
+		return errors.New("engine already registered with splitter")
+	}
+
+	e.engines[engine] = struct{}{}
+
+	return nil
+}
+
+func (e *Engine) UnregisterEngine(engine module.Engine) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	delete(e.engines, engine)
+}
+
+// Ready returns a ready channel that is closed once the engine has fully
+// started. For the splitter engine, this is true once all of the
+// registered engines have started.
+func (e *Engine) Ready() <-chan struct{} {
+	return e.unit.Ready(func() {
+		for engine := range e.engines {
+			<-engine.Ready()
+		}
+	})
+}
+
+// Done returns a done channel that is closed once the engine has fully stopped.
+// For the splitter engine, this is true once all of the registered engines
+// have stopped.
+func (e *Engine) Done() <-chan struct{} {
+	return e.unit.Done(func() {
+		for engine := range e.engines {
+			<-engine.Done()
+		}
+	})
+}
+
+// SubmitLocal submits an event originating on the local node.
+func (e *Engine) SubmitLocal(event interface{}) {
+	e.unit.Launch(func() {
+		err := e.ProcessLocal(event)
+		if err != nil {
+			engine.LogError(e.log, err)
+		}
+	})
+}
+
+// Submit submits the given event from the node with the given origin ID
+// for processing in a non-blocking manner. It returns instantly and logs
```

Comment:
Better to add a test verifies that sending 10 messages concurrently, will end up each engine receiving 10 messages.

Reply:
Okay, I can add a test.